Add a functional test for SLURM (Docker'ized)

kinow opened the pull request:
These changes partially address #3372 

Adds a `dockerfiles` folder (same as in JupyterHub) where we can store `Dockerfile`'s for things like functional tests with SLURM, PBS, SSH, etc.

Changes the `.travis.yml` file to load the Docker services, and to build the image locally, install Cylc Flow that was checked out (volume), and runs a new functional test that executes a simple workflow that uses the SLURM batch system handler.

In my fork, it took Travis ~2 minutes to a) fetch the layers for the container, b) build the container locally, c) install `cylc-flow` with `pip`, and d) run the functional test. This was much faster than I expected.

This is just an initial example to show what I had in mind with that issue for CI running batch systems. Would be better if there was an easier way to isolate the SLURM functional tests. I didn't know how to do that, so used an environment variable (quite sure I saw that in another test, skipping based on env vars), or if we could already write the test with `pytest`.

But hopefully this should allow us to start testing more parts of Cylc Flow that were not tested as they require things like SLURM server, PBS server, SSH connectivity, etc.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

kinow commented:
This was my Friday fun-project, while reading the Cylc User Guide, and experimenting with SLURM (only used PBS). Wasn't sure if that would work, so finished the work yesterday, but only confirmed Travis tests passed this morning.

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=5177696&bid=16993946)

kinow commented:
Re. the Codacy issues, the `ubuntu:latest` is not the latest as in other images. For the Ubuntu image, it's the LTS version.

![image](https://user-images.githubusercontent.com/304786/76659996-c4f7b500-65dc-11ea-8eb6-7f2082f0e63d.png)

The other two warnings about pinning the version in `apt-get` and using a parameter are not necessary as that LTS version doesn't get lots of updates, and we have few dependencies.

kinow commented on dockerfiles/slurm/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+VOLUME ["/root/cylc-flow"]
+
+ENV LC_ALL=C.UTF-8
+ENV LANG=C.UTF-8
```
Without the above the image fails and says something like the Cylc module needing the above locale, and that it doesn't work with ASCII I think (should have noted the message)

kinow commented on dockerfiles/slurm/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+#  - https://stackoverflow.com/questions/46966876/installing-emulating-slurm-on-an-ubuntu-16-04-desktop-slurmd-fails-to-start
+
+RUN apt-get update && apt-get install -y \
+    lsof \
```
The test battery uses this to look at running jobs I think. The container failed locally when I wrote the functional test and complained that it couldn't find `lsof`.

oliver-sanders replied:
FYI

Cylc is supposed to be able to run on POSIX systems provided GNU coreutils are installed.

The test battery, however, depends on a few more GNU utils, including `sed`, `awk` and apparently `lsof`.

kinow commented on dockerfiles/slurm/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+
+ENV LC_ALL=C.UTF-8
+ENV LANG=C.UTF-8
+ENV USER=root
```
Docker doesn't set this, but it's used by Cylc to locate the workflows I think. Added after it failed locally.

oliver-sanders replied:
I thought `USER` was one of those dependable variables which should always be there.

kinow commented on tests/slurm/00-basic/suite.rc:
```diff
@@ -0,0 +1,9 @@
+[[hello]]
+script = "sleep 2; echo Hello World!"
+[[[job]]]
+batch system = slurm
```
From an example in the CUG.

oliver-sanders commented:
Woohoo, the beginnings of our little docker swarm!

BTW we have a head start in testing because we already have some slurm tests in the functional test battery ATM which are skipped if slurm is not provided e.g. tests/cylc-poll/08-slurm.t

kinow commented:
> Woohoo, the beginnings of our little docker swarm!
> 
> BTW we have a head start in testing because we already have some slurm tests in the functional test battery ATM which are skipped if slurm is not provided e.g. tests/cylc-poll/08-slurm.t

Thanks! Should have looked for tests there.

Just had a look to see if I could include it, the only way I found to run it was with the following diff.

```diff
diff --git a/tests/cylc-poll/06-loadleveler.t b/tests/cylc-poll/06-loadleveler.t
index 61649d268..e2777db67 100755
--- a/tests/cylc-poll/06-loadleveler.t
+++ b/tests/cylc-poll/06-loadleveler.t
@@ -16,23 +16,14 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #-------------------------------------------------------------------------------
 # Test "cylc poll" for loadleveler, slurm, or pbs jobs.
-export CYLC_TEST_IS_GENERIC=false
 . "$(dirname "$0")/test_header"
 #-------------------------------------------------------------------------------
-BATCH_SYS_NAME="${TEST_NAME_BASE##??-}"
-RC_PREF="[test battery][batch systems][$BATCH_SYS_NAME]"
-CYLC_TEST_BATCH_TASK_HOST=$( \
-    cylc get-global-config -i "${RC_PREF}host" 2>'/dev/null')
-CYLC_TEST_BATCH_SITE_DIRECTIVES=$( \
-    cylc get-global-config -i "${RC_PREF}[directives]" 2>'/dev/null')
-if [[ -z "${CYLC_TEST_BATCH_TASK_HOST}" || "${CYLC_TEST_BATCH_TASK_HOST}" == None ]]
+if [[ -z "${CYLC_TEST_BATCH_SLURM+x}" || -z "${CYLC_TEST_BATCH_SLURM}" || "${CYLC_TEST_BATCH_SLURM}" == None ]]
 then
-    skip_all "\"[test battery][batch systems][$BATCH_SYS_NAME]host\" not defined"
+    skip_all "no slurm installation"
 fi
-export CYLC_TEST_BATCH_TASK_HOST CYLC_TEST_BATCH_SITE_DIRECTIVES
+#-------------------------------------------------------------------------------
 set_test_number 2
+#-------------------------------------------------------------------------------
 reftest
-if [[ "${CYLC_TEST_BATCH_TASK_HOST}" != 'localhost' ]]; then
-    purge_suite_remote "${CYLC_TEST_BATCH_TASK_HOST}" "${SUITE_NAME}"
-fi
 exit
```

i.e. used the same variable I created to control whether the test is executed or not, and ignored the remote part.

I didn't know about the `CYLC_TEST_IS_GENERIC` variable, but looks like I should use it instead of creating a new one, or use both? My idea is to allow for running PBS tests, SLURM tests, SSH tests, isolated. As a container wouldn't necessarily have all the configurations for every batch system, so we would need to run groups of tests per batch system.

I also executed the test above with `docker exec -e CYLC_TEST_BATCH_SLURM=true -w /root/cylc-flow slurm ./etc/bin/run-functional-tests.sh -v tests/cylc-poll/08-slurm.t` which gives:

(It fails locally due to missing SLURM, `sbatch`, etc.)

```bash
[01:00:50] tests/cylc-poll/08-slurm.t .. 
1..2
REGISTERED cylctb-20200316T010016Z/cylc-poll/08-slurm -> /tmp/tmp.vTh7rqEcE6/cylctb-20200316T010016Z/cylc-poll/08-slurm
ok 1 - 08-slurm-validate
ok 2 - 08-slurm-run
ok    31323 ms ( 0.00 usr  0.00 sys +  9.49 cusr  0.88 csys = 10.37 CPU)
[01:00:50]
All tests successful.
Files=1, Tests=2, 33 wallclock secs ( 0.02 usr  0.00 sys +  9.49 cusr  0.88 csys = 10.39 CPU)
Result: PASS
```

But I think we also need remote tests. Initially the Docker image had both, SLURM and `openssh`, with a user `root` with password `root` (should be OK as this is a container only for tests).

But I had tried something similar in the past, and from what I can remember, I couldn't get Cylc to work with user:pass? I think it expects to have SSH keys correctly set up.

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=5177696&bid=16993946)

hjoliver commented:
Very nice @kinow. I guess we could do the same with the non-"pro" version of PBS.  (Although you don't have to do that on this PR :-)

oliver-sanders commented:
> the only way I found to run it was with the following diff

Try running it via `etc/bin/run-functional-tests`, there are environment variables which get provided by this script and also `tests/lib/bash/test_header`.

kinow commented:
> 
> 
> Very nice @kinow. I guess we could do the same with the non-"pro" version of PBS. (Although you don't have to do that on this PR :-)

Yup, that's the plan. Just adding an initial container and some simple tests, then we can have a few more iterations and improve this work and add more tests/platforms.

This same technique can be used for remote hosts too with SSH.

oliver-sanders commented:
Hopefully this will be an easy port to GitHub actions...

hjoliver commented:
Current Slurm tests in the test battery:
```
git grep 'batch system' tests flakytests/ | grep slurm
tests/cylc-get-site-config/05-host-bool-override.t:        batch system = slurm
tests/cylc-poll/08-slurm/suite.rc:            batch system = slurm
tests/directives/slurm/suite.rc:            batch system = slurm
tests/execution-time-limit/02-slurm.t:RC_PREF="[test battery][batch systems][$CYLC_TEST_BATCH_SYS]"
tests/execution-time-limit/02-slurm.t:    skip_all "\"[test battery][batch systems][$CYLC_TEST_BATCH_SYS]host\" not defined"
tests/execution-time-limit/02-slurm/suite.rc:            batch system = {{CYLC_TEST_BATCH_SYS}}
tests/job-kill/03-slurm/suite.rc:            batch system=slurm
tests/jobscript/06-slurm-no-directives/suite.rc:            batch system = slurm
```

`CYLC_TEST_IS_GENERIC` is documented in `/etc/bin/run-functional-tests.sh`:
```
Platform-specific tests must set "CYLC_TEST_IS_GENERIC=false" before sourcing the test_header.
```
So I think a generic test is one that runs a local suite with local background jobs - i.e. you can run it anywhere.

For local test battery runs with PBS or Slurm etc. we set a hosts etc. in the test global config file.  From `/cylc/flow/cfgspec/globalcfg.py`:
```
   'test battery': {
        'remote host with shared fs': [VDR.V_STRING],
        'remote host': [VDR.V_STRING],
        'remote owner': [VDR.V_STRING],
        'batch systems': {
            '__MANY__': {
                'host': [VDR.V_STRING],
                'out viewer': [VDR.V_STRING],
                'err viewer': [VDR.V_STRING],
                'directives': {
                    '__MANY__': [VDR.V_STRING],
                },  
            },  
        },  
    },  
```

kinow commented:
>For local test battery runs with PBS or Slurm etc. we set a hosts etc. in the test global config file. From /cylc/flow/cfgspec/globalcfg.py

Ah, ok. So if I remove my newly created environment variable, then mark the test as not-generic, and add the docker hostname to this `/cylc/flow/cfgspec/globalcfg.py` file, then it should be executed only when running on that hostname?

hjoliver commented:
> then it should be executed only when running on that hostname? 

If a host name is defined for Slurm tests, the slurm tests will be executed and the associated suites will submit their jobs to Slurm on the Slurm host.

kinow commented:
Updated the Docker container and now it goes with a `flow-tests.rc` file. @hjoliver I think a couple of the tests you found were actually generic ones.

```bash
kinow@ranma:~/Development/python/workspace/cylc-flow$ docker exec -e CYLC_TEST_RUN_GENERIC=false -w /root/cylc-flow slurm ./etc/bin/run-functional-tests.sh -v tests/cylc-get-site-config/05-host-bool-override.t tests/cylc-poll/08-slurm.t tests/directives/02-slurm.t tests/execution-time-limit/02-slurm.t tests/job-kill/03-slurm.t tests/jobscript/06-slurm-no-directives.t
[02:40:04] tests/cylc-get-site-config/05-host-bool-override.t .. skipped: not running generic tests
[02:40:04] tests/jobscript/06-slurm-no-directives.t ............ skipped: not running generic tests
[02:40:54] tests/cylc-poll/08-slurm.t .......................... 
1..2
REGISTERED cylctb-20200408T024004Z/cylc-poll/08-slurm -> /tmp/tmp.KKltk9g833/cylctb-20200408T024004Z/cylc-poll/08-slurm
ok 1 - 08-slurm-validate
ok 2 - 08-slurm-run
ok    47025 ms ( 0.00 usr  0.00 sys +  7.78 cusr  0.78 csys =  8.56 CPU)
[02:40:58] tests/job-kill/03-slurm.t ........................... 
1..2
REGISTERED cylctb-20200408T024004Z/job-kill/03-slurm -> /tmp/tmp.VnIhxZaTSL/cylctb-20200408T024004Z/job-kill/03-slurm
ok 1 - 03-slurm-validate
ok 2 - 03-slurm-run
ok    51793 ms ( 0.00 usr  0.00 sys + 15.66 cusr  1.56 csys = 17.22 CPU)
[02:43:38] tests/execution-time-limit/02-slurm.t ............... 
1..3
REGISTERED cylctb-20200408T024004Z/execution-time-limit/02-slurm -> /tmp/tmp.CbarQegoN9/cylctb-20200408T024004Z/execution-time-limit/02-slurm
ok 1 - 02-slurm-validate
ok 2 - 02-slurm-run
ok 3 - job-grep-ok
ok   211605 ms ( 0.00 usr  0.00 sys + 24.32 cusr  2.40 csys = 26.72 CPU)
[02:43:46] tests/directives/02-slurm.t ......................... 
1..2
REGISTERED cylctb-20200408T024004Z/directives/02-slurm -> /tmp/tmp.N5fLskISUA/cylctb-20200408T024004Z/directives/02-slurm
ok 1 - 02-slurm-validate
ok 2 - 02-slurm-run
ok   219229 ms ( 0.00 usr  0.00 sys + 33.17 cusr  3.29 csys = 36.46 CPU)
[02:43:46]
All tests successful.
Files=6, Tests=9, 222 wallclock secs ( 0.03 usr  0.01 sys + 33.20 cusr  3.31 csys = 36.55 CPU)
Result: PASS
```

Updated my two new functional tests too. Running **without** Docker:

```bash
(venv) kinow@ranma:~/Development/python/workspace/cylc-flow$ ./etc/bin/run-functional-tests.sh -v tests/slurm/
skipped: "[test battery][batch systems][slurm]host" not defined
[16:02:01] tests/slurm/01-special-characters.t .. skipped: "[test battery][batch systems][slurm]host" not defined
[16:02:01]
Files=2, Tests=0,  2 wallclock secs ( 0.02 usr  0.00 sys +  2.76 cusr  0.19 csys =  2.97 CPU)
Result: NOTESTS
```

And running with Docker.

```bash
kinow@ranma:~/Development/python/workspace/cylc-flow$ docker exec -e CYLC_TEST_RUN_GENERIC=false -w /root/cylc-flow slurm ./etc/bin/run-functional-tests.sh -v ./tests/slurm/
[02:48:39] ./tests/slurm/00-basic.t ............... 
1..1
REGISTERED cylctb-20200408T024827Z/slurm/00-basic -> /tmp/tmp.MNqUjV57ye/cylctb-20200408T024827Z/slurm/00-basic
ok 1 - 00-basic-run
ok    10103 ms ( 0.00 usr  0.00 sys +  5.19 cusr  0.48 csys =  5.67 CPU)
[02:48:42] ./tests/slurm/01-special-characters.t .. 
1..1
REGISTERED cylctb-20200408T024827Z/slurm/01-special-characters -> /tmp/tmp.PWtai34L48/cylctb-20200408T024827Z/slurm/01-special-characters
ok 1 - 01-special-characters-run
ok    13094 ms ( 0.00 usr  0.00 sys + 10.25 cusr  1.06 csys = 11.31 CPU)
[02:48:42]
All tests successful.
Files=2, Tests=2, 14 wallclock secs ( 0.02 usr  0.00 sys + 10.25 cusr  1.06 csys = 11.33 CPU)
Result: PASS
```

kinow commented:
Now up to Travis to say whether it all works or not :sweat_smile:

kinow commented:
Conflicts resolved locally.

When running non-generic tests, the `./tests/parsec` tests always fail, unless you have `python3` correctly set up in the environment. Fixed that locally too already, by changing from 3.6 to python 3.7 in Ubuntu (inside the container).

Now just final tweaks as `./cylc/cylc-submit` has a couple tests failing. One of the (03-at) runs as non-generic, but requires `at` module. Instead of changing the module, I will just include `atd` in the container.

kinow commented:
Travis now happy, and SLURM tests passing (both new and existing non-generic ones)

oliver-sanders commented on dockerfiles/slurm/slurm.submit:
```diff
@@ -0,0 +1,13 @@
+date
+hostname
+sleep 5
+date
```
What's the purpose of this file?

kinow replied:
the script used to submit a sample job. Prints time, hostname, sleeps, then prints date again. I guess it could just print hostname or something and avoid sleeping perhaps? (quite sure I copied it from some blog/doc)

oliver-sanders replied:
I guess we should cut the sleep out at least.

kinow replied:
Agreed, good catch! Will update in the morning.

oliver-sanders commented on dockerfiles/slurm/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+    supervisor \
+    python3.7 \
+    python3-pip \
+    python3.7-dev \
```
Installing both 3.7 and 3.7-dev?

kinow replied:
It was missing headers for... pyzmq or protobuf? I think of of these modules failed. Initially I installed only python3.7. Then noticed it was missing pip, then added it. And kept adding modules as the docker image failed to successfully build & run

kinow commented:
>I thought USER was one of those dependable variables which should always be there.

Me too, but not in the base image. I got docker images (especially those using alpine) that didn't have basic variables, or utilities I thought would always be present in a linux installation.

kinow commented:
@oliver-sanders even if approved, we can delay merging it after the GitHub actions in :+1: (I haven't used GitHub actions with Docker yet, so looking forward to trying that :grimacing: )

oliver-sanders commented:
Would be good to combine this with GH actions caching to avoid re-building the docker image for each build.

kinow commented:
(Pushed a new commit just to remove the `sleep` from the example job script)

oliver-sanders commented:
Curiously that upset Travis.

kinow commented:
Yup, failed on my fork too. No idea why. Couldn't find after quickly searching the log and re-reading last change. Weird. Will have another look tomorrow, sorry.

kinow commented:
I have confirmed that the `slurm.submit` script is included with the Docker image, but **not** used anywhere. That's just an example that can be used by whoever builds the container to confirm that slurm is working.

Interestingly, after pushing a commit to add the second `date` command back, it worked on my fork's Travis CI build,

![image](https://user-images.githubusercontent.com/304786/79526474-79dc3080-80b9-11ea-97c8-cae851f28495.png)

But failed again on cylc/cylc-flow. It seems to have run all four tests. But one of the two `poll_grep` executed afterwards probably failed.

![image](https://user-images.githubusercontent.com/304786/79526597-bb6cdb80-80b9-11ea-89b1-d2337d169db6.png)

This test appears to be a bit flaky. Should we move to the flakyTests group and leave it disabled?

I executed it twice locally with the Docker image and using `cylc-submit` directory/group it passed fine (i.e. `docker exec -e CYLC_TEST_RUN_GENERIC=false -w /root/cylc-flow slurm ./etc/bin/run-functional-tests.sh -v ./tests/cylc-submit`). But it failed when I executed all tests (i.e. `... ./tests/`).

kinow commented:
(close-open to re-trigger Travis for this branch)

TomekTrzeciak commented:
Is the current thinking to eventually have several containers, each for a different batch system, or would it make sense to build one container with several batch systems in it? They should be non-conflicting and tests could be possibly the same, just parameterised with different submit directives.

kinow commented:
> Is the current thinking to eventually have several containers, each for a different batch system, or would it make sense to build one container with several batch systems in it? They should be non-conflicting and tests could be possibly the same, just parameterised with different submit directives.

Good question @TomekTrzeciak . I don't have a strong preference for either a single container or multiple containers. While one gives you a single image to maintain, and probably faster build time, it may also give you problems further ahead in case the batch systems conflict with each other, or affect the execution of tests in other batch systems.

I had a bit of experience with PBS Torque on CentOS from a previous project, and learned that SLURM on Ubuntu is super simple to set up. Never tried both in the same environment. Maybe somebody else with better knowledge could weigh in.

I'd be happy to try helping to implement either scenario (both would be interesting to set up, and increase the scenarios of our tests :nerd_face: so nerdly-interesting)

oliver-sanders commented:
Travis skips a lot of tests, the roles we need to fill are:

* Remote host
* Remote host with shared filesystem
* Slurm
* PBS
* loadleveler
* ...

I would have thought that it might be difficult to install all of the supported batch systems in a single image but I've never tried.

The first two will require separate images, I think we could do worse than having a different docker image for each purpose, at least it keeps the images simple, this one is already quite long.

oliver-sanders commented:
>> Is the current thinking to eventually have several containers, each for a different batch system

> I don't have a strong preference for either a single container or multiple containers

Looks like we never got around to making an umbrella issue for our CI testing requirements so I've distilled that here - https://github.com/cylc/cylc-flow/issues/3641

The lack of remote testing is really hitting us, I've spent quite a few hours in the past days fixing remote tests 😠.

We've got a lot of remote-type work coming up shortly so if there is a quick win it would be great to have it!

kinow commented:
Rebased

kinow commented:
This is probably superseded by new Docker environments for functional testing, that also uses platforms. Closing :+1:

hjoliver commented:
@kinow we don't actually have Slurm platforms configured in the new docker swarm yet.  It would be good to take what you've learned about installing and configuring Slurm in a container here, and apply that there. (Or do you have that on the back burner already @oliver-sanders?)

oliver-sanders commented:
I haven't added slurm etc yet, however, we now have a framework to do so and slurm tests will run as soon as we add a compatible platform to the swarm.

Just need to build a clurm image that extends the cylc-remote image (for TCP comms) and add the following to `etc/conf/global.cylc`:

```
[platforms]
    [[_local_slurm_indep_tcp]]
        batch system = slurm
```
